### DISCUSSION TITLE
Integrating a segmentation model in 3D Slicer

### DISCUSSION THREAD

**Discussion by D123Roy** [Project-MONAI/MONAILabel#1094] [General]
My final goal is to use interactive segmentation on a pre-defined model with monailabel on 3D Slicer. I have generated a model (a .pth file) and now trying to integrate it into 3D Slicer by adding the path (in config deepedit.py) of .pth to the model.
I have tested the model and it is generating fine results. However, on running the model in 3D Slicer-auto segmentation section it gives completely off the track labels to Spleen.

I referred to -
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/spleen_segmentation_3d.ipynb

to generate the weights for the model.

My question is how do we integrate these weight into the 3D Slicer to follow the ultimate guide for interactive segmentation? Please suggest ! Do we need to make some changes to the infer folder as well ?

**Comment by tangy5** [comment]
Hey, @D123Roy glad to see you are here as well. 
In case you haven't seen my reply in the other repo, I paste the reply here: please see below


if you are starting a monai label server with the radiology app, you could do this:

See here, if no model provided, the app will initialize and download a pretrained model from website:
https://github.com/Project-MONAI/MONAILabel/blob/739226b381caa1a6c4579633a06e19f9bdf04178/sample-apps/radiology/lib/configs/segmentation_spleen.py#L59

To use your own model, you could put your model in the model folder name f"pretrained_{name}.pt" , for example: pretrained_segmentation_spleen.pt "", add a conf argument: --conf use_pretrained_model false.

Then the monai label server should be able to load the model for auto segmentation.


In addition, the MONAI Label itself can do interactive segmentation without extra that tutorial, you could see how active learning is performed referring to: https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology

For you case, if you wouldn't like to use the default settings such as infer transforms, you could customized this script for Slicer:
https://github.com/Project-MONAI/MONAILabel/blob/739226b381caa1a6c4579633a06e19f9bdf04178/sample-apps/radiology/lib/infers/segmentation_spleen.py#L55

You can also modify Training paradigms similarly. Then use them in the Slicer UI.

**Reply by D123Roy** [reply]
Thanks I am grateful for all the support I received from you !
Actually I can see my custom pre-trained model (.pth file) being loaded in cmd.
But not generating desired results with auto-segmentation modules. (I trained the module with Spleen data)

**Reply by tangy5** [reply]
There might be some transform inconsistency between your code for MONAI Label app and the way you developed using tutorial/customized code. 
You can check transforms here, if you are using the "segmentation_spleen" model:

https://github.com/Project-MONAI/MONAILabel/blob/20608cc81922f6a3027fa25bde9c5964cb4dd74f/sample-apps/radiology/lib/infers/segmentation_spleen.py#L55

**Comment by D123Roy** [comment]
Another issue that I want  to add is that - On adding my pre-trained weights in "segmentation/segmentation_spleen" model, the system is unable to load the model and hangs when I try to run the auto-segmentation module. Hence I have been using "Deepedit model" in running purposes. I hope this wouldnt be an issue for generating such results.

**Reply by tangy5** [reply]
There might be issues with the code when customizing segmentation_spleen, it would be better to help if you can provide more details of the implementation. Thanks.

**Comment by ebaldelomar** [comment]
Hi, I have a tangent question that I thought maybe would be okay to follow up in this thread. Is it possible to take an existing open source model a colleague has developed:

pth files: https://drive.google.com/drive/folders/1AhvNaoB7delzgznm4dLkkvZmGipCQShd
based of this software: https://github.com/hrlblab/Glo-In-One/tree/main

I'm curious if it's possible to just run this model through 3D Slicer (which I love using, though still new to learning MONAI and auto segmentation)?? Any guidance is much appreciated!
